// @companion-module/base/dist/module-api/base.js
import { createModuleLogger } from '../logging.js';
import { assertNever } from '../util.js';
import { SharedUdpSocketImpl, } from './shared-udp-socket.js';
import { isInstanceContext } from '../host-api/context.js';
export class InstanceBase {
    #context;
    #logger;
    #options;
    get id() {
        return this.#context.id;
    }
    get instanceOptions() {
        return this.#options;
    }
    /**
     * The user chosen name for this instance.
     * This can be changed just before `configUpdated` is called
     */
    get label() {
        return this.#context.label;
    }
    /**
     * Create an instance of the module
     */
    constructor(internal) {
        if (!isInstanceContext(internal) || !internal._isInstanceContext)
            throw new Error(`Module instance is being constructed incorrectly. Make sure you aren't trying to do this manually`);
        this.#context = internal;
        this.#logger = createModuleLogger();
        this.createSharedUdpSocket = this.createSharedUdpSocket.bind(this);
        this.#options = {
            disableVariableValidation: false,
            disableNewConfigLayout: false,
        };
        this.log('debug', 'Initializing');
    }
    saveConfig(newConfig, newSecrets) {
        this.#context.saveConfig(newConfig, newSecrets);
    }
    /**
     * Set the action definitions for this instance
     * @param actions The action definitions
     */
    setActionDefinitions(actions) {
        this.#context.setActionDefinitions(actions);
    }
    /**
     * Set the feedback definitions for this instance
     * @param feedbacks The feedback definitions
     */
    setFeedbackDefinitions(feedbacks) {
        this.#context.setFeedbackDefinitions(feedbacks);
    }
    /**
     * Set the preset definitions for this instance
     * @param structure The structure of the preset sections
     * @param presets The unstructured preset definitions
     */
    setPresetDefinitions(structure, presets) {
        this.#context.setPresetDefinitions(structure, presets);
    }
    /**
     * Set the variable definitions for this instance
     * @param variables The variable definitions
     */
    setVariableDefinitions(variables) {
        if (Array.isArray(variables))
            throw new Error('Variable definitions should be an object, not an array');
        this.#context.setVariableDefinitions(variables);
    }
    /**
     * Set the values of some variables
     * @param values The new values for the variables
     */
    setVariableValues(values) {
        this.#context.setVariableValues(values);
    }
    /**
     * Get the last set value of a variable from this connection
     * @param variableId id of the variable
     * @returns The value
     */
    getVariableValue(variableId) {
        return this.#context.getVariableValue(variableId);
    }
    /**
     * Request all feedbacks of all types to be re-executed
     */
    checkAllFeedbacks() {
        this.#context.checkAllFeedbacks();
    }
    /**
     * Request all feedbacks of the specified types to be re-executed.
     * At least one feedback type must be provided.
     * @param feedbackType The first feedback type to check
     * @param feedbackTypes Additional feedback types to check
     */
    checkFeedbacks(feedbackType, ...feedbackTypes) {
        this.#context.checkFeedbacks([feedbackType, ...feedbackTypes]);
    }
    /**
     * Request the specified feedback instances to be be re-executed
     * @param feedbackIds The ids of the feedback instances to check
     */
    checkFeedbacksById(...feedbackIds) {
        this.#context.checkFeedbacksById(feedbackIds);
    }
    /**
     * Call subscribe on all currently known placed actions.
     * It can be useful to trigger this upon establishing a connection, to ensure all data is loaded.
     * @param actionIds The actionIds to call subscribe for. If no values are provided, then all are called.
     */
    subscribeActions(...actionIds) {
        this.#context.subscribeActions(actionIds);
    }
    /**
     * Call unsubscribe on all currently known placed actions.
     * It can be useful to do some cleanup upon a connection closing.
     * @param actionIds The actionIds to call subscribe for. If no values are provided, then all are called.
     */
    unsubscribeActions(...actionIds) {
        this.#context.unsubscribeActions(actionIds);
    }
    /**
     * Call unsubscribe on all currently known placed feedbacks.
     * It can be useful to do some cleanup upon a connection closing.
     * @param feedbackIds The feedbackIds to call subscribe for. If no values are provided, then all are called.
     */
    unsubscribeFeedbacks(...feedbackIds) {
        this.#context.unsubscribeFeedbacks(feedbackIds);
    }
    /**
     * Add an action to the current recording session
     * @param action The action to be added to the recording session
     * @param uniquenessId A unique id for the action being recorded. This should be different for each action, but by passing the same as a previous call will replace the previous value.
     */
    recordAction(action, uniquenessId) {
        this.#context.recordAction(action, uniquenessId);
    }
    /**
     * Send an osc message from the system osc sender
     * @param host destination ip address
     * @param port destination port number
     * @param path message path
     * @param args message arguments
     */
    oscSend(host, port, path, args) {
        this.#context.oscSend(host, port, path, args);
    }
    /**
     * Update the status of this connection
     * @param status The status level
     * @param message Additional information about the status
     *
     * ### Example
     * ```js
     * this.updateStatus(InstanceStatus.Ok)
     * ```
     */
    updateStatus(status, message) {
        this.#context.updateStatus(status, message ?? null);
    }
    /**
     * Write a line to the log
     * @param level The level of the message
     * @param message The message text to write
     */
    log(level, message) {
        switch (level) {
            case 'debug':
                this.#logger.debug(message);
                break;
            case 'info':
                this.#logger.info(message);
                break;
            case 'warn':
                this.#logger.warn(message);
                break;
            case 'error':
                this.#logger.error(message);
                break;
            default:
                assertNever(level);
                this.#logger.info(message);
                break;
        }
    }
    createSharedUdpSocket(typeOrOptions, callback) {
        const options = typeof typeOrOptions === 'string' ? { type: typeOrOptions } : typeOrOptions;
        const socket = new SharedUdpSocketImpl(this.#context, options);
        if (callback)
            socket.on('message', callback);
        return socket;
    }
}
