// @companion-module/base/dist/module-api/upgrade.js
/**
 * A helper upgrade script, which does nothing.
 * Useful to replace a script which is no longer needed
 */
export const EmptyUpgradeScript = () => ({
    updatedConfig: null,
    updatedSecrets: null,
    updatedActions: [],
    updatedFeedbacks: [],
});
/**
 * A helper script to automate the bulk of the process to upgrade feedbacks from 'advanced' to 'boolean'.
 * There are some built-in rules for properties names based on the most common cases.
 * @param upgradeMap The feedbacks to upgrade and the properties to convert
 */
export function CreateConvertToBooleanFeedbackUpgradeScript(upgradeMap) {
    // Warning: the unused parameters will often be null
    return (_context, props) => {
        const changedFeedbacks = [];
        for (const feedback of props.feedbacks) {
            let upgrade_rules = upgradeMap[feedback.feedbackId];
            if (upgrade_rules === true) {
                // These are some automated built in rules. They can help make it easier to migrate
                upgrade_rules = {
                    bg: 'bgcolor',
                    bgcolor: 'bgcolor',
                    fg: 'color',
                    color: 'color',
                    png64: 'png64',
                    png: 'png64',
                };
            }
            if (upgrade_rules) {
                if (!feedback.style)
                    feedback.style = {};
                for (const [option_key, style_key] of Object.entries(upgrade_rules)) {
                    const rawVal = feedback.options[option_key];
                    if (rawVal !== undefined) {
                        feedback.style[style_key] = rawVal;
                        delete feedback.options[option_key];
                        changedFeedbacks.push(feedback);
                    }
                }
            }
        }
        return {
            updatedConfig: null,
            updatedSecrets: null,
            updatedActions: [],
            updatedFeedbacks: changedFeedbacks,
        };
    };
}
/**
 * A helper script to automate the bulk of the process to upgrade feedbacks from having a module defined 'invert' field, to use the builtin one.
 * The feedback definitions must be updated manually, this can only help update existing usages of the feedback.
 * @param upgradeMap The feedbacks to upgrade and the id of the option to convert
 */
export function CreateUseBuiltinInvertForFeedbacksUpgradeScript(upgradeMap) {
    // Warning: the unused parameters will often be null
    return (_context, props) => {
        const changedFeedbacks = [];
        for (const feedback of props.feedbacks) {
            const propertyName = upgradeMap[feedback.feedbackId];
            if (typeof propertyName !== 'string')
                continue;
            // Retrieve and delete the old value
            const rawValue = feedback.options[propertyName];
            if (rawValue === undefined)
                continue;
            delete feedback.options[propertyName];
            // Interpret it to a boolean, it could be stored in a few ways
            if (!rawValue.isExpression) {
                feedback.isInverted = {
                    isExpression: false,
                    value: rawValue.value === 'true' || Boolean(rawValue.value) === true || Number(rawValue.value) > 0,
                };
            }
            else {
                // We can't fix this case for them
                feedback.isInverted = rawValue;
            }
            changedFeedbacks.push(feedback);
        }
        return {
            updatedConfig: null,
            updatedSecrets: null,
            updatedActions: [],
            updatedFeedbacks: changedFeedbacks,
        };
    };
}
/**
 * Helper to fix up values which could be a number or a string of variables into the new expression format
 * This has some logic to try to preserve the existing behaviour as much as possible, while also keeping the output user friendly
 * @param value The value to fix up
 * @returns The fixed up value, or the original value if it did not need to be changed
 */
export function FixupNumericOrVariablesValueToExpressions(value) {
    // Nothing to do if already an expression!
    if (!value || value.isExpression)
        return value;
    const oldValNum = Number(value.value);
    const oldValRaw = value.value;
    if (typeof value.value === 'number' ||
        (typeof value.value === 'string' && value.value.trim() !== '' && !isNaN(oldValNum))) {
        // It looks like a plain number, so store it as one
        return {
            isExpression: false,
            value: oldValNum,
        };
    }
    else if (typeof oldValRaw === 'string') {
        const trimmedStr = oldValRaw.trim();
        // Crudely check if it looks like a simple variable
        if (trimmedStr.startsWith('$(') && trimmedStr.endsWith(')') && !trimmedStr.slice(2).includes('$(')) {
            // It does, so we can treat that as a plain expression!
            return {
                isExpression: true,
                value: oldValRaw,
            };
        }
        else {
            // Otherwise its something more complex, so wrap it in a parseVariables and hope for the best!
            return {
                isExpression: true,
                value: `parseVariables("${oldValRaw.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}")`,
            };
        }
    }
    else {
        // Its not a string, or anything vaguely sane..
        // Nothing really we can do
        return value;
    }
}
/**
 * Helper to fix up values which could be a boolean, number, or a string of variables into the new expression format.
 * Numbers are coerced to booleans (0 = false, non-zero = true).
 * The string literals "true" and "false" (case-insensitive) are also coerced to booleans.
 * @param value The value to fix up
 * @returns The fixed up value, or the original value if it did not need to be changed
 */
export function FixupBooleanOrVariablesValueToExpressions(value) {
    // Nothing to do if already an expression!
    if (!value || value.isExpression)
        return value;
    const oldValRaw = value.value;
    if (typeof oldValRaw === 'boolean') {
        // Already a boolean, nothing to do
        return value;
    }
    else if (typeof oldValRaw === 'number') {
        // Coerce number to boolean
        return {
            isExpression: false,
            value: oldValRaw !== 0,
        };
    }
    else if (typeof oldValRaw === 'string') {
        const trimmedStr = oldValRaw.trim();
        const lowerStr = trimmedStr.toLowerCase();
        if (lowerStr === 'true' || lowerStr === 'false') {
            // Coerce string literal to boolean
            return {
                isExpression: false,
                value: lowerStr === 'true',
            };
        }
        else if (!isNaN(Number(trimmedStr)) && trimmedStr !== '') {
            // Coerce numeric string to boolean
            return {
                isExpression: false,
                value: Number(trimmedStr) !== 0,
            };
        }
        else if (trimmedStr.startsWith('$(') && trimmedStr.endsWith(')') && !trimmedStr.slice(2).includes('$(')) {
            // It looks like a simple variable, treat as a plain expression
            return {
                isExpression: true,
                value: oldValRaw,
            };
        }
        else {
            // Something more complex, wrap it in a parseVariables and hope for the best!
            return {
                isExpression: true,
                value: `bool(parseVariables("${oldValRaw.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"))`,
            };
        }
    }
    else {
        // Not a string, boolean, or number — nothing we can do
        return value;
    }
}
